File: ProgrammingProjects/python/Project_03/main.py
"""Load a Project 3 case and prepare an RHF calculation."""

from pathlib import Path

from chemistry.electronic_structure import (
    RHFCalculator,
    SCFOptions,
    load_ao_integrals,
)


INPUT_DIRECTORY = (
    Path(__file__).resolve().parents[2]
    / "resources"
    / "Project_03"
    / "input"
    / "h2o"
    / "STO-3G"
)


def main(input_directory: Path = INPUT_DIRECTORY) -> None:
    integrals = load_ao_integrals(input_directory)
    calculator = RHFCalculator(
        integrals,
        SCFOptions(
            energy_tolerance=1.0e-10,
            density_tolerance=1.0e-8,
            max_iterations=100,
        ),
    )

    print(f"Input directory: {input_directory}")
    print(f"Basis functions: {integrals.n_basis}")
    print(f"Electrons: {integrals.n_electrons}")
    print(f"Occupied RHF orbitals: {calculator.n_occupied}")
    print(
        "Nuclear repulsion: "
        f"{integrals.nuclear_repulsion:.12f} Eh"
    )
    result = calculator.run()

    print()
    print("SCF iterations")
    print(
        f"{'Iter':>4}"
        f" {'E(elec) / Eh':>20}"
        f" {'E(total) / Eh':>20}"
        f" {'Delta E / Eh':>14}"
        f" {'Delta D':>14}"
    )
    print("-" * 78)

    for iteration in result.history:
        print(
            f"{iteration.index:4d}"
            f" {iteration.electronic_energy:20.12f}"
            f" {iteration.total_energy:20.12f}"
            f" {iteration.energy_change:14.6e}"
            f" {iteration.density_change:14.6e}"
        )

    print()
    print(f"Converged: {result.converged}")
    print(f"SCF iterations: {result.history[-1].index}")
    print(
        "Final electronic energy: "
        f"{result.electronic_energy:.12f} Eh"
    )
    print(
        "Nuclear repulsion: "
        f"{integrals.nuclear_repulsion:.12f} Eh"
    )
    print(f"Final total energy: {result.total_energy:.12f} Eh")


if __name__ == "__main__":
    main()

File: ProgrammingProjects/cpp/Project_03/main.cpp
#include "minihf/integrals.h"
#include "minihf/rhf_calculator.h"

#include <exception>
#include <filesystem>
#include <iomanip>
#include <iostream>

int main(int argc, char* argv[])
{
    if (argc != 2) {
        std::cerr << "Usage: project03 <Project-3-input-directory>\n";
        return 1;
    }

    try {
        const std::filesystem::path input_directory{argv[1]};
        const minihf::AOIntegrals integrals =
            minihf::AOIntegrals::from_directory(input_directory);
        const minihf::RHFCalculator calculator{integrals};

        std::cout
            << "Input directory: " << input_directory.string() << '\n'
            << "Basis functions: " << integrals.basis_size() << '\n'
            << "Electrons: " << integrals.electron_count() << '\n'
            << "Occupied RHF orbitals: "
            << calculator.occupied_orbitals() << '\n'
            << std::fixed << std::setprecision(12)
            << "Nuclear repulsion: "
            << integrals.nuclear_repulsion() << " Eh\n";

        const minihf::RHFResult result = calculator.run();

        std::cout
            << "\nSCF iterations\n"
            << std::setw(4) << "Iter"
            << ' ' << std::setw(20) << "E(elec) / Eh"
            << ' ' << std::setw(20) << "E(total) / Eh"
            << ' ' << std::setw(14) << "Delta E / Eh"
            << ' ' << std::setw(14) << "Delta D" << '\n'
            << "------------------------------------------------------------------------------\n";

        for (const minihf::SCFIteration& iteration : result.history) {
            std::cout
                << std::setw(4) << iteration.index
                << ' ' << std::fixed << std::setprecision(12)
                << std::setw(20) << iteration.electronic_energy
                << ' ' << std::setw(20) << iteration.total_energy
                << ' ' << std::scientific << std::setprecision(6)
                << std::setw(14) << iteration.energy_change
                << ' ' << std::setw(14) << iteration.density_change
                << '\n';
        }

        std::cout
            << "\nConverged: "
            << (result.converged ? "True" : "False") << '\n'
            << "SCF iterations: " << result.iterations() << '\n'
            << std::fixed << std::setprecision(12)
            << "Final electronic energy: "
            << result.electronic_energy << " Eh\n"
            << "Nuclear repulsion: "
            << integrals.nuclear_repulsion() << " Eh\n"
            << "Final total energy: "
            << result.total_energy << " Eh\n";

    }
    catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }

    return 0;
}
